### PR TITLE
Issue/8

### LINKED ISSUE #8: Customizing CodeSystem properties heirarchyMeaning and jurisdiction are unsupported.  Setting contacts does not add them to the transform output.

### PR DESCRIPTION
**Pull request by alecdorenkamp**
See Issue/8.

Fixed contacts not being created properly.
Made heirarchyMeaning configurable.
Made jurisdiction configurable.

Added E2E tests to Validate that CodeSystem metadata added in this issue are being represented properly.
    Please see the small refactor in how the program exits in Application.java.
    If this pattern is acceptable, it will be easy to add tests for the other fields in the future using the same testing approach.

Fixed Display for base Thing being represented as the full Thing uri instead of simply "Thing"

### CODE REVIEW

File: src/main/java/au/csiro/fhir/owl/Application.java [hunk 1]
```diff
@@ -163,6 +163,10 @@ public void run(String... args) {
     options.addOption("experimental", false, "Indicates if the code system is for testing "
         + "purposes or real usage.");
     
+    options.addOption("heirarchyMeaning", true, "The meaning of the hierarchy of concepts as "
```
Comment (jimsteel) [inline src/main/java/au/csiro/fhir/owl/Application.java]:
```suggestion
    options.addOption("hierarchyMeaning", true, "The meaning of the hierarchy of concepts as "
```

File: README.md [hunk 2]
```diff
@@ -67,10 +67,12 @@ The following options are available:
...
 | -help              | none        | Print the help message. |
+| -heirarchyMeaning  | string      | The meaning of the hierarchy of concepts as represented in this resource. Valid values are *grouped-by*, *is-a*, *part-of*, and *classified-with*.  Default is *is-a*. | 
 | -i                 | string      | The input OWL file. |
 | -id                | string      | The technical id of the code system. Required if using PUT to upload the resource to a FHIR server. |
-| -identifier        | string      | Comma-separated list of additional business identifiers. Each business identifer has the format [system]|[value]. |
+| -identifier        | string      | Comma-separated list of additional business identifiers. Each business identifer has the format [system]\|[value]. |
 | -includeDeprecated | boolean     | Include all OWL classes, including deprecated ones. |
+| -jurisdiction      | string      | Comma-separated list of jurisdictions for the codesystem. Each jurisdiction must have the format [code\|system\|display], with values retrieved from the [FHIR Jurisdiction ValueSet](https://hl7.org/fhir/valueset-jurisdiction.html). |
```
Comment (jimsteel) [inline README.md]: system|code|display might be more in keeping with FHIR's canonical syntax (which has system before code)

File: src/test/java/au/csiro/fhir/owl/util/ArgConstants.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    public static final String EXAMPLE_EMAIL = "lorem.ipsum@gmail.com";
+    public static final String EMAIL = "email";
+    public static final String EXAMPLE_NAME_2 = "Ms. Ipsum Lorem";
+    public static final String EXAMPLE_PHONE = "987-765-4321";
+    public static final String PHONE = "phone";
+    
+    public static final String CONTACT_EXAMPLE_WITH_EMAIL = EXAMPLE_NAME_1 +"|"+EMAIL+"|"+ EXAMPLE_EMAIL;
+    public static final String CONTACT_EXAMPLE_WITH_PHONE = EXAMPLE_NAME_2 +"|"+PHONE+"|"+ EXAMPLE_PHONE;
+    
+    public static final String HEIRARCHY_MEANING_FLAG = "-heirarchyMeaning";
```
Comment (jimsteel) [inline src/test/java/au/csiro/fhir/owl/util/ArgConstants.java]:
```suggestion
    public static final String HIERARCHY_MEANING_FLAG = "-hierarchyMeaning";
```

File: src/test/java/au/csiro/fhir/owl/util/ArgConstants.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    public static final String HEIRARCHY_MEANING_GROUPED_BY = "grouped-by"; //Static value of CodeSystem.CodeSystemHierarchyMeaning.GROUPEDBY
```
Comment (jimsteel) [inline src/test/java/au/csiro/fhir/owl/util/ArgConstants.java]:
```suggestion
    public static final String HIERARCHY_MEANING_GROUPED_BY = "grouped-by"; //Static value of CodeSystem.CodeSystemHierarchyMeaning.GROUPEDBY
```

File: src/test/java/au/csiro/fhir/owl/MetadataFromInputTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import static au.csiro.fhir.owl.util.ArgConstants.CONTACT_EXAMPLE_WITH_EMAIL;
+import static au.csiro.fhir.owl.util.ArgConstants.CONTACT_EXAMPLE_WITH_PHONE;
+import static au.csiro.fhir.owl.util.ArgConstants.CONTACT_FLAG;
+import static au.csiro.fhir.owl.util.ArgConstants.EMAIL;
+import static au.csiro.fhir.owl.util.ArgConstants.EXAMPLE_EMAIL;
+import static au.csiro.fhir.owl.util.ArgConstants.EXAMPLE_NAME_1;
+import static au.csiro.fhir.owl.util.ArgConstants.EXAMPLE_NAME_2;
+import static au.csiro.fhir.owl.util.ArgConstants.EXAMPLE_PHONE;
+import static au.csiro.fhir.owl.util.ArgConstants.HEIRARCHY_MEANING_FLAG;
+import static au.csiro.fhir.owl.util.ArgConstants.HEIRARCHY_MEANING_GROUPED_BY;
```
Comment (jimsteel) [inline src/test/java/au/csiro/fhir/owl/MetadataFromInputTest.java]:
```suggestion
import static au.csiro.fhir.owl.util.ArgConstants.HIERARCHY_MEANING_FLAG;
import static au.csiro.fhir.owl.util.ArgConstants.HIERARCHY_MEANING_GROUPED_BY;
```

File: src/test/java/au/csiro/fhir/owl/MetadataFromInputTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+/**
+ * E2E tests for fhir-owl that validate that CodeSystem metadata values are properly represented when given as input.
+ */
+
+@SpringBootTest(args = {
+    TEST_FLAG,
+    INPUT_FLAG, INPUT_FILE, 
+    OUTPUT_FLAG, OUTPUT_FILE,
+    HEIRARCHY_MEANING_FLAG, HEIRARCHY_MEANING_GROUPED_BY,
```
Comment (jimsteel) [inline src/test/java/au/csiro/fhir/owl/MetadataFromInputTest.java]:
```suggestion
    HIERARCHY_MEANING_FLAG, HIERARCHY_MEANING_GROUPED_BY,
```

File: README.md [hunk 7]
```diff
@@ -67,10 +67,12 @@ The following options are available:
 | -descriptionProp   | string      | Comma-separated list of OWL annotation properties that contain the code system description. |
 | -experimental      | boolean     | Indicates if the code system is for testing purposes or real usage. |
 | -help              | none        | Print the help message. |
+| -heirarchyMeaning  | string      | The meaning of the hierarchy of concepts as represented in this resource. Valid values are *grouped-by*, *is-a*, *part-of*, and *classified-with*.  Default is *is-a*. |
```
Comment (jimsteel) [inline README.md]:
```suggestion
| -hierarchyMeaning  | string      | The meaning of the hierarchy of concepts as represented in this resource. Valid values are *grouped-by*, *is-a*, *part-of*, and *classified-with*.  Default is *is-a*. | 
```

File: src/main/java/au/csiro/fhir/owl/FhirOwlService.java [hunk 8]
```diff
@@ -446,9 +453,14 @@ CodeSystem createCodeSystem(
...
     final String valueset = csp.getValueSet();
     cs.setValueSet(Objects.requireNonNullElseGet(valueset, () -> createVsUrl(cs.getUrl())));
-    
-    // Hierarchy meaning - always is-a for OWL ontologies
-    cs.setHierarchyMeaning(CodeSystemHierarchyMeaning.ISA);
+  
+    // HeirarchyMeaning
+    final String heirarchyMeaning = csp.getHeirarchyMeaning();
+    if (heirarchyMeaning != null) {
+      cs.setHierarchyMeaning(CodeSystem.CodeSystemHierarchyMeaning.fromCode(heirarchyMeaning));
```
Comment (jimsteel) [inline src/main/java/au/csiro/fhir/owl/FhirOwlService.java]:
```suggestion
    final String hierarchyMeaning = csp.getHierarchyMeaning();
    if (hierarchyMeaning != null) {
      cs.setHierarchyMeaning(CodeSystem.CodeSystemHierarchyMeaning.fromCode(hierarchyMeaning));
```

File: src/main/java/au/csiro/fhir/owl/FhirOwlService.java [hunk 9]
```diff
@@ -446,9 +453,14 @@ CodeSystem createCodeSystem(
     // Value set
     final String valueset = csp.getValueSet();
     cs.setValueSet(Objects.requireNonNullElseGet(valueset, () -> createVsUrl(cs.getUrl())));
-    
-    // Hierarchy meaning - always is-a for OWL ontologies
-    cs.setHierarchyMeaning(CodeSystemHierarchyMeaning.ISA);
+  
+    // HeirarchyMeaning
```
Comment (jimsteel) [inline src/main/java/au/csiro/fhir/owl/FhirOwlService.java]:
```suggestion
    // HierarchyMeaning
```

File: src/main/java/au/csiro/fhir/owl/Application.java [hunk 10]
```diff
@@ -411,12 +428,22 @@ private CodeSystemProperties loadCodeSystemProperties(CommandLine line) {
...
+    }
 
     val = line.getOptionValue("purpose");
     if (val != null) {
       res.setPurpose(val);
     }
     
+    val = line.getOptionValue("heirarchyMeaning");
+    if (val != null) {
+      res.setHeirarchyMeaning(val);
```
Comment (jimsteel) [inline src/main/java/au/csiro/fhir/owl/Application.java]:
```suggestion
    val = line.getOptionValue("hierarchyMeaning");
    if (val != null) {
      res.setHierarchyMeaning(val);
```

File: src/main/java/au/csiro/fhir/owl/CodeSystemProperties.java [hunk 11]
```diff
@@ -41,9 +44,11 @@ public class CodeSystemProperties extends OwlProperties {
   private String publisher = null;
   private final List<String> publisherProps = new ArrayList<>();
   private final List<ContactDetail> contacts = new ArrayList<>();
+  private final List<CodeableConcept> jurisdictions = new ArrayList<>();
   private String description = null;
   private final List<String> descriptionProps = new ArrayList<>();
   private String purpose = null;
+  private String heirarchyMeaning = null;
```
Comment (jimsteel) [inline src/main/java/au/csiro/fhir/owl/CodeSystemProperties.java]:
```suggestion
  private String hierarchyMeaning = null;
```

File: src/main/java/au/csiro/fhir/owl/CodeSystemProperties.java [hunk 12]
```diff
@@ -457,6 +499,24 @@ public String getPurpose() {
...
+    return heirarchyMeaning;
+  }
+  
+  /**
+   * Sets the heirarchyMeaning.
+   *
+   * @param heirarchyMeaning the heirarchyMeaning to set
+   */
+  public void setHeirarchyMeaning(String heirarchyMeaning) {
+    this.heirarchyMeaning = heirarchyMeaning;
```
Comment (jimsteel) [inline src/main/java/au/csiro/fhir/owl/CodeSystemProperties.java]:
```suggestion
   * Returns the hierarchyMeaning.
   *
   * @return the hierarchyMeaning
   */
  public String getHierarchyMeaning() {
    return hierarchyMeaning;
  }
  
  /**
   * Sets the hierarchyMeaning.
   *
   * @param hierarchyMeaning the hierarchyMeaning to set
   */
  public void setHierarchyMeaning(String hierarchyMeaning) {
    this.hierarchyMeaning = hierarchyMeaning;
```

**Review by jimsteel — CHANGES_REQUESTED** [review]
Some typos first

**Comment by alecdorenkamp** [comment]
Thanks for the feedback @jimsteel!  I believe I've made all the requested changes (hierarchy spelling and changing arg order)

**Review by jimsteel — APPROVED** [review]

**Pull request merged**